src/Microsoft.OData.Edm/ExtensionMethods/AnnotationsTargetHelper.cs

```diff
@@ -0,0 +1,205 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="AnnotationsTargetHelper.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.Edm
+{
+    using System.Collections.Generic;
+    using Microsoft.OData.Edm.Vocabularies;
+
+    /// <summary>
+    /// AnnotationsTarget helper
+    /// </summary>
+    internal static class AnnotationsTargetHelper
+    {
+        public static IEdmVocabularyAnnotatable HandleTargetSegments(IEdmModel model, List<IEdmElement> results, string[] targetSegments, int targetSegmentsCount, int level)
+        {
+            string elementName = targetSegments[0];
+
+            IEdmEntityContainer container = model.FindEntityContainer(elementName);
+
+            if (container != null)
+            {
+                results.Add(container);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmEntityContainer(model, results, container, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    return container;
+                }
+            }
+
+            return null;
+        }
+
+        public static IEdmVocabularyAnnotatable HandleIEdmEntityContainer(IEdmModel model, List<IEdmElement> results, IEdmEntityContainer edmEntityContainer, string[] targetSegments, int targetSegmentsCount, int level)
+        {
+            IEdmEntitySet entitySet = edmEntityContainer.FindEntitySetExtended(targetSegments[level]);
+
+            if (entitySet != null)
+            {
+                results.Add(entitySet);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmContainerElement(model, results, entitySet, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    return entitySet;
+                }
+            }
+
+            IEdmSingleton singleton = edmEntityContainer.FindSingletonExtended(targetSegments[level]);
+
+            if (singleton != null)
+            {
+                results.Add(singleton);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmContainerElement(model, results, singleton, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    return singleton;
+                }
+            }
+
+            return null;
+        }
+
+        public static IEdmVocabularyAnnotatable HandleIEdmContainerElement(IEdmModel model, List<IEdmElement> results, IEdmEntityContainerElement edmEntityContainerElement, string[] targetSegments, int targetSegmentsCount, int level)

```

Comment:
Refer to my other comment https://github.com/OData/odata.net/pull/2769/files#r1382848419